src/Compilers/Core/Portable/IReferenceOrISignature.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.Symbols;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Used to devirtualize Dictionary/HashSet for EqualityComparer{T}.Default
+    /// </summary>
+    internal struct IReferenceOrISignatureEquivalent : IEquatable<IReferenceOrISignatureEquivalent>
```

Comment:
I'm not sure I understand why we would need both of these: it seems like just `IReferenceOrISignatureEquivalent should be fine.

Reply:
`IReferenceOrISignature` uses `ReferenceEquals` for equals and `IReferenceOrISignatureEquivalent` has a more complex equals, previously provided by `MetadataEntityReferenceComparer.cs`

The dictionary, hashtable and concurrentdictionary will devirtualize (and potentially inline) the equals for a struct when the Comparer is unspecified (if it also implements `IEquatable`); whereas specifying a Comparer will be a slower interface call.

Reply:
This would be great info to have in a comment in the file itself :)

---
In reply to: [480276103](https://github.com/dotnet/roslyn/pull/47274#discussion_r480276103) [](ancestors = 480276103)